File: csharp/src/GuardianForge.Core/AnomalyDetector.cs
namespace GuardianForge.Core;

/// <summary>
/// The statistical/pattern anomaly layer: per-agent detection of tool loops, rate spikes,
/// and cyclic A-B-A-B patterns. Deterministic and cheap; runs on every event before any
/// model is consulted.
/// </summary>
public sealed class AnomalyDetector
{
    private readonly object _lock = new();
    private readonly Dictionary<string, List<string>> _history = new();
    private const int Window = 12;
    private const int LoopThreshold = 5;
    private const int RateThreshold = 10;

    private static string Fp(AgentEvent ev) => $"{ev.Type}:{ev.Tool}:{ev.Action}";

    /// <summary>Record an event and return any anomalies detected for the agent.</summary>
    public List<Anomaly> Observe(AgentEvent ev)
    {
        lock (_lock)
        {
            if (!_history.TryGetValue(ev.AgentId, out var h))
            {
                h = new List<string>();
                _history[ev.AgentId] = h;
            }
            h.Add(Fp(ev));
            if (h.Count > Window) h.RemoveRange(0, h.Count - Window);

            var outList = new List<Anomaly>();
            if (DetectLoop(ev.AgentId, h) is { } loop) outList.Add(loop);
            if (DetectRate(ev.AgentId, h) is { } rate) outList.Add(rate);
            if (DetectCyclic(ev.AgentId, h) is { } cyc) outList.Add(cyc);
            return outList;
        }
    }

    private static Anomaly? DetectLoop(string agent, List<string> h)
    {
        if (h.Count == 0) return null;
        var last = h[^1];
        var count = 0;
        for (var i = h.Count - 1; i >= 0 && h[i] == last; i--) count++;
        if (count >= LoopThreshold)
        {
            var score = Round(Math.Min(1.0, (double)count / Window));
            return new Anomaly { AgentId = agent, Type = AnomalyType.Loop, Score = score, Detail = $"{count} consecutive identical actions ({last})" };
        }
        return null;
    }

    private static Anomaly? DetectRate(string agent, List<string> h)
    {
        if (h.Count >= RateThreshold)
        {
            var score = Round(Math.Min(1.0, (double)h.Count / Window));
            return new Anomaly { AgentId = agent, Type = AnomalyType.RateSpike, Score = score, Detail = $"{h.Count} events within the observation window" };
        }
        return null;
    }

    private static Anomaly? DetectCyclic(string agent, List<string> h)
    {
        if (h.Count < 6) return null;
        var tail = h.GetRange(h.Count - 6, 6);
        var a = tail[0];
        var b = tail[1];
        if (a == b) return null;
        for (var i = 0; i < tail.Count; i++)
        {
            var want = i % 2 == 0 ? a : b;
            if (tail[i] != want) return null;
        }
        return new Anomaly { AgentId = agent, Type = AnomalyType.Cyclic, Score = 0.7, Detail = $"cyclic pattern between \"{a}\" and \"{b}\"" };
    }

    private static double Round(double f) => Math.Round(f, 2);
}

File: java/src/main/java/com/guardianforge/AnomalyDetector.java
package com.guardianforge;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * The deterministic pattern-anomaly layer.
 *
 * <p>Mirrors the Go {@code anomaly} package: it keeps a bounded per-agent history and flags
 * loops (the same action hammered over and over), rate spikes, and cyclic A-B-A-B patterns.
 * It runs on every event before any LLM is consulted.
 */
public final class AnomalyDetector {
    private final Map<String, List<String>> history = new HashMap<>();
    private final int window = 12;
    private final int loopThreshold = 5;
    private final int rateThreshold = 10;

    /** Build a detector with the reference defaults. */
    public AnomalyDetector() {
    }

    private static String fp(AgentEvent ev) {
        return ev.type + ":" + ev.tool + ":" + ev.action;
    }

    /**
     * Record an event and return any anomalies detected for its agent.
     *
     * @param ev the event to observe
     * @return the anomalies flagged (loop, then rate, then cyclic order preserved)
     */
    public List<Anomaly> observe(AgentEvent ev) {
        List<String> h = history.getOrDefault(ev.agentId, new ArrayList<>());
        h = new ArrayList<>(h);
        h.add(fp(ev));
        if (h.size() > window) {
            h = new ArrayList<>(h.subList(h.size() - window, h.size()));
        }
        history.put(ev.agentId, h);

        List<Anomaly> out = new ArrayList<>();
        Anomaly loop = detectLoop(ev.agentId, h);
        if (loop != null) {
            out.add(loop);
        }
        Anomaly rate = detectRate(ev.agentId, h);
        if (rate != null) {
            out.add(rate);
        }
        Anomaly cyclic = detectCyclic(ev.agentId, h);
        if (cyclic != null) {
            out.add(cyclic);
        }
        return out;
    }

    private Anomaly detectLoop(String agent, List<String> h) {
        if (h.isEmpty()) {
            return null;
        }
        String last = h.get(h.size() - 1);
        int count = 0;
        for (int i = h.size() - 1; i >= 0 && h.get(i).equals(last); i--) {
            count++;
        }
        if (count >= loopThreshold) {
            double score = clamp((double) count / (double) window);
            return new Anomaly(agent, AnomalyType.LOOP, round(score),
                    count + " consecutive identical actions (" + last + ")");
        }
        return null;
    }

    private Anomaly detectRate(String agent, List<String> h) {
        if (h.size() >= rateThreshold) {
            double score = clamp((double) h.size() / (double) window);
            return new Anomaly(agent, AnomalyType.RATE_SPIKE, round(score),
                    h.size() + " events within the observation window");
        }
        return null;
    }

    private Anomaly detectCyclic(String agent, List<String> h) {
        if (h.size() < 6) {
            return null;
        }
        List<String> tail = h.subList(h.size() - 6, h.size());
        String a = tail.get(0);
        String b = tail.get(1);
        if (a.equals(b)) {
            return null;
        }
        for (int i = 0; i < tail.size(); i++) {
            String want = (i % 2 == 1) ? b : a;
            if (!tail.get(i).equals(want)) {
                return null;
            }
        }
        return new Anomaly(agent, AnomalyType.CYCLIC, 0.7,
                "cyclic pattern between \"" + a + "\" and \"" + b + "\"");
    }

    private static double clamp(double f) {
        if (f > 1) {
            return 1;
        }
        if (f < 0) {
            return 0;
        }
        return f;
    }

    private static double round(double f) {
        return (double) ((int) (f * 100 + 0.5)) / 100;
    }
}
